### ISSUE TITLE
Error deleting CloudFront Response Headers Policy

### ISSUE THREAD

**Issue by ZsoltPath** [hashicorp/terraform-provider-aws#21730]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version
Terraform v0.15.1
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v3.64.2

### Affected Resource(s)

* aws_cloudfront_response_headers_policy
* aws_cloudfront_distribution

### Terraform Configuration Files

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

```hcl
resource "aws_cloudfront_distribution" "site" {

  wait_for_deployment = false

  origin {
    domain_name = aws_s3_bucket.site[0].bucket_domain_name 
    origin_id   = "origin-${aws_s3_bucket.site[0].id}"
       s3_origin_config {
      origin_access_identity = aws_cloudfront_origin_access_identity.site[0].cloudfront_access_identity_path
    }
  }

  custom_error_response {
    error_caching_min_ttl = 300
    error_code            = 404
    response_code         = 200
    response_page_path    = "/index.html"
  }

  enabled             = true
  price_class         = "PriceClass_100"
  retain_on_delete    = false
  default_root_object = "index.html"
  web_acl_id          = var.waf_web_acl_id

  default_cache_behavior {
    allowed_methods  = ["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"]
    cached_methods   = ["GET", "HEAD", "OPTIONS"]
    target_origin_id = "origin-${aws_s3_bucket.site[0].id}"
    compress         = true
    response_headers_policy_id = aws_cloudfront_response_headers_policy.site.id

    forwarded_values {
      query_string = false
      headers      = []

      cookies {
        forward = "none"
      }
    }

    viewer_protocol_policy = "redirect-to-https"
    min_ttl                = 0
    default_ttl            = 0
    max_ttl                = 86400
  }

  viewer_certificate {
    acm_certificate_arn      = var.certificates[each.key]
    ssl_support_method       = "sni-only"
    minimum_protocol_version = "TLSv1.2_2019"
  }

  restrictions {
    geo_restriction {
      restriction_type = "none"
    }
  }


  tags = var.tags

}

resource "aws_cloudfront_response_headers_policy" "site" {
  name    = "${local.environment}-${var.site_name}"

  security_headers_config  {
    content_security_policy {
      content_security_policy ="default-src 'none';"
      override = true
    }
}

```

### Expected Behavior

I was trying to remove the response header resource from the config.

The plan quite rightly shows it

```
# module.admin.aws_cloudfront_distribution.site will be updated in-place
  ~ resource "aws_cloudfront_distribution" "site" {
        id                             = "E2U9TRDNITAK52"
        tags                           = {
            "service" = "admin"
        }
        # (19 unchanged attributes hidden)


      ~ default_cache_behavior {
          - response_headers_policy_id = "9b2e0ed0-f4bb-4d9d-ab3a-8f9f848b835d" -> null
            # (11 unchanged attributes hidden)

        }



        # (4 unchanged blocks hidden)
    }

  # module.admin.aws_cloudfront_response_headers_policy.site will be destroyed
  - resource "aws_cloudfront_response_headers_policy" "site" {
      - etag = "E1PA6795UKMFR9" -> null
      - id   = "9b2e0ed0-f4bb-4d9d-ab3a-8f9f848b835d" -> null
      - name = "dev-admin" -> null

      - security_headers_config {
          - content_security_policy {
              - content_security_policy = "default-src 'none';" -> null
              - override                = true -> null
            }
        }
    }
```

I'm expecting the resource destroyed and the link from the distribution removed.

### Actual Behavior

Getting an error message 

```
│ Error: error deleting CloudFront Response Headers Policy (9b2e0ed0-f4bb-4d9d-ab3a-8f9f848b835d): ResponseHeadersPolicyInUse: The specified response headers policy is currently associated with a cache behavior. Please disassociate the policy before deleting.
│       status code: 409, request id: 3ec0ff73-1df8-43b8-a893-781cdb4c92d6
```

I guess it's a chicken and egg problem and I'm not sure if possible to handle in terraform. 
The response header link should be removed from the distribution first and then destroy the policy. But the terraform dependency works the other way around.

I think it could be the same with the aws_cloudfront_origin_request_policy resource

### Steps to Reproduce

1. create aws_cloudfront_response_headers_policy resource and add it to a cloudfront distribution
2. `terraform apply`
3. remove the resource from the config
4. `terraform apply`

**Comment by albertoCaroM** [comment]
I had the same problem this morning, and I had fixed  it in two steps.


First  of all you have to delete the asignation 

    **response_headers_policy_id = aws_cloudfront_response_headers_policy.site.id**
 
but you have to keep the resource  "aws_cloudfront_response_headers_policy" in your terraform configuration file.

Them  run the "terraform apply"

 Edit again the configuration and delete the resource  "aws_cloudfront_response_headers_policy" .
For last  execute again "terrafom apply"

**Comment by ZsoltPath** [comment]
@albertoCaroM I know the workaround, but in a CI environment it's impossible.
We're creating and destroying test environments multiple times a day. I can't manually hack each of them.

**Comment by dsech** [comment]
Same issue happens with `aws_cloudfront_function` associations. 

```
│ Error: error deleting CloudFront Function (frontend_response): FunctionInUse: Cannot delete function frontend_response, it is in use by 1 distributions
│       status code: 409, request id: d4605f97-5993-4425-aa77-109cd732a01f
|
```

First the removal of the `function_association` block inside the distribution must be applied, and only after that the removal of the `aws_cloudfront_function` resource can be applied.

**Comment by noahlz** [comment]
Another work-around: ClickOps 

Go to the AWS Console and unassociate the Headers Response Policy from the Distribution behavior.  Even though your terraform state still thinks its associated, it at least circumvents this error and allows terraform to both update the cloudfront distribution to use the new response policy (I went from a custom one to a managed one, in this case), and delete the unneeded policy.

(This doesn't help those with CI/CD constraints, but got me moving again. YMMV).

**Comment by briananstett** [comment]
This is a really annoying problem. It's too bad CloudFront won't accept a CloudFront Function's ARN before the function is published. We could have the function depend on the CloudFront distribution and just assume the function's ARN for the association.

```terraform
    function_association {
        event_type   = "viewer-request"
        function_arn = "arn:aws:cloudfront::${data.aws_caller_identity.this.account_id}:function/${local.name}"
    }
    
    
resource "aws_cloudfront_function" "this" {
  name    = local.name
  runtime = "cloudfront-js-1.0"
  publish = true
  code    = templatefile("${path.module}/src/auth.js")

  depends_on = [
    aws_cloudfront_distribution.this
  ]
}
```

```bash
InvalidArgument: The parameter FunctionAssociationArn is invalid; dev-4444-4-2-test-assets was not found or is not published.
```

**Comment by a-nych** [comment]
This is still valid in the `v4.6.0` version.

One cannot remove the `aws_cloudfront_cache_policy` or `aws_cloudfront_response_headers_policy` without targeting the CloudFront resource first, which destroys the flow of CI/CD.

**Comment by nick-kostov** [comment]
I have noticed that when I want to attach the policy in CF distribution I am not being able to.

I refer to it but still when I apply it it is detaching it.



`resource "aws_cloudfront_distribution" "this" {
  depends_on = [
    aws_cloudfront_response_headers_policy.security
  ]
  comment = "Dev Distribution"
  enabled             = true
  is_ipv6_enabled     = false
  default_root_object = "index.html"
  aliases             = [format("*.%s", local.domain_name), format("www.%s", local.domain_name), local.domain_name]
  price_class         = var.cf.price_class

  dynamic "origin" {
    for_each = toset(var.cf.origin_ids)
    content {
      domain_name = aws_s3_bucket.this.bucket_domain_name
      origin_id   = origin.key
      s3_origin_config {
        origin_access_identity = aws_cloudfront_origin_access_identity.this.cloudfront_access_identity_path
      }
    }
  }

  restrictions {
    geo_restriction {
      restriction_type = "none"
    }
  }

  default_cache_behavior {
    target_origin_id = var.cf.origin_ids[0]

    allowed_methods = ["GET", "HEAD", "OPTIONS"]
    cached_methods  = ["GET", "HEAD"]

    # Either this or the origin/cf cache policies below
    # forwarded_values {
    #   query_string = false
    #   cookies {
    #     forward = "none"
    #   }
    # }

    viewer_protocol_policy = "redirect-to-https"
    min_ttl                = 0
    default_ttl            = 7200
    max_ttl                = 86400

    origin_request_policy_id = local.origin_cache_policy
    cache_policy_id          = local.cf_cache_policy
    response_headers_policy_id = aws_cloudfront_response_headers_policy.security.id

  }

  dynamic "ordered_cache_behavior" {
    for_each = length(var.cf.origin_ids) > 1 ? local.ordered_origin_ids : []
    content {
      path_pattern     = "${ordered_cache_behavior.value}/*"
      target_origin_id = ordered_cache_behavior.value

      allowed_methods        = ["GET", "HEAD", "OPTIONS"]
      cached_methods         = ["GET", "HEAD"]
      compress               = true
      viewer_protocol_policy = "redirect-to-https"
      min_ttl                = 0
      default_ttl            = 7200
      max_ttl                = 86400

      # Caching policies
      origin_request_policy_id = local.origin_cache_policy
      cache_policy_id          = local.cf_cache_policy
      response_headers_policy_id = aws_cloudfront_response_headers_policy.security.id
    }
  }

  viewer_certificate {
    acm_certificate_arn = local.ssl_cert_arn
    ssl_support_method  = "sni-only"
  }

  # Custom error response
  custom_error_response {
    error_caching_min_ttl = 0
    error_code            = 404
    response_code         = 200
    response_page_path    = "/"
  }

  # Ignore changes done by CircleCI
  lifecycle {
    ignore_changes = [origin, default_cache_behavior, ordered_cache_behavior]
  }
}`

**Comment by JeremieBethmont** [comment]
This issue seems to be still happening with AWS provider `v4.22.0` when replacing a `aws_cloudfront_response_headers_policy` resource.

```
│ Error: error deleting CloudFront Response Headers Policy (78b01a03-76dc-4618-9f99-e084a09c0091): ResponseHeadersPolicyInUse: The specified response headers policy is currently associated with a cache behavior. Please disassociate the policy before deleting.
│ 	status code: 409, request id: edc15669-6f03-47af-8b41-a4e5de212a51
```

The workaround is to go and detach the policy manually before running the terraform apply.

**Comment by yurios** [comment]
Issues is still reproducible in the `5.47.0`

**Comment by jar-b** [comment]
Hello everyone and thanks for your engagement with this issue.

TLDR; For this scenario, a [`create_before_destroy` lifecycle argument](https://developer.hashicorp.com/terraform/language/meta-arguments/lifecycle#create_before_destroy) on the policy resource allows the graph ordering to be modified when the policy is destroyed. Specifically, this switches the order of operations so that the distribution update occurs _before_ the destruction of the policy. Assuming the policy is attached to only a single distribution, it can then be safely deleted.

More context is available in the Terraform core documentation on resource destruction.
- https://github.com/hashicorp/terraform/blob/v1.15.8/docs/destroying.md#create-before-destroy

---

Here is a minimal example configuration.

<details>
  <summary>Show/Hide Configuration</summary>

```hcl
# Step 1 - Create distribution and policy together.
# Step 2 - Remove policy and reference in distribution resource.
# 
#   Note: The lifecycle block can be commented out to observe the default
#   order of operations during destroy (policy deletion, then distribution
#   update). By adding the create_before_destroy argument, the order is
#   inverted and the distribution update completes in its entirety before
#   policy destruction begins.
#
# Relates https://github.com/hashicorp/terraform-provider-aws/issues/21730

terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 6.0"
    }
  }
}

# Configure the AWS Provider
provider "aws" {}

resource "aws_cloudfront_response_headers_policy" "test" {
  name = "jb-test"

  security_headers_config {
    content_security_policy {
      content_security_policy = "default-src 'none';"
      override                = true
    }
  }

  # Required to invert the order of operations during destroy
  lifecycle {
    create_before_destroy = true
  }
}

resource "aws_cloudfront_distribution" "test" {
  enabled          = false
  retain_on_delete = false

  origin {
    domain_name = "www.example.com"
    origin_id   = "test"

    custom_origin_config {
      http_port              = 80
      https_port             = 443
      origin_protocol_policy = "https-only"
      origin_ssl_protocols   = ["TLSv1.2"]
    }
  }

  default_cache_behavior {
    allowed_methods        = ["GET", "HEAD"]
    cached_methods         = ["GET", "HEAD"]
    target_origin_id       = "test"
    viewer_protocol_policy = "allow-all"
    response_headers_policy_id = aws_cloudfront_response_headers_policy.test.id

    forwarded_values {
      query_string = false

      cookies {
        forward = "none"
      }
    }
  }

  restrictions {
    geo_restriction {
      restriction_type = "none"
    }
  }

  viewer_certificate {
    cloudfront_default_certificate = true
  }
}
```

</details>

With the above applied, it is possible to observe the behavior during policy removal both with and without the lifecycle argument.

Without a `create_before_destroy` lifecycle argument, the default order of operations is to delete the policy and then update the distribution. This triggers the failure reported in the original issue.

```console
% terraform apply
...

Plan: 0 to add, 1 to change, 1 to destroy.
aws_cloudfront_response_headers_policy.test: Destroying... [id=4bc3ce47-75f1-4104-82af-9fefabec4d87]
╷
│ Error: deleting CloudFront Response Headers Policy (4bc3ce47-75f1-4104-82af-9fefabec4d87): operation error CloudFront: DeleteResponseHeadersPolicy, https response error StatusCode: 409, RequestID: 4396421b-200e-4923-b409-d7fe3bb5b1f9, ResponseHeadersPolicyInUse: The specified response headers policy is currently associated with a cache behavior. Please disassociate the policy before deleting.
```

Adding the `create_before_destroy` lifecycle argument causes the order of operations to be inverted, meaning the distribution update will complete in its entirety before policy deletion begins.

> [!NOTE]
> When adding/uncommenting the `lifecycle` block, be sure to run a `terraform apply` with the policy (and lifecycle block) still in configuration. This ensures the `create_before_destroy` value is written to state and accounted for on the following `apply` where the policy is removed and distribution is updated.

```console
% terraform apply
...

Plan: 0 to add, 1 to change, 1 to destroy.
aws_cloudfront_distribution.test: Modifying... [id=E2GSZ0BU4OH406]
aws_cloudfront_distribution.test: Still modifying... [id=E2GSZ0BU4OH406, 00m10s elapsed]
aws_cloudfront_distribution.test: Still modifying... [id=E2GSZ0BU4OH406, 00m20s elapsed]
aws_cloudfront_distribution.test: Still modifying... [id=E2GSZ0BU4OH406, 00m30s elapsed]
aws_cloudfront_distribution.test: Still modifying... [id=E2GSZ0BU4OH406, 00m40s elapsed]
aws_cloudfront_distribution.test: Still modifying... [id=E2GSZ0BU4OH406, 00m50s elapsed]
aws_cloudfront_distribution.test: Modifications complete after 51s [id=E2GSZ0BU4OH406]
aws_cloudfront_response_headers_policy.test: Destroying... [id=4bc3ce47-75f1-4104-82af-9fefabec4d87]
aws_cloudfront_response_headers_policy.test: Destruction complete after 0s

Apply complete! Resources: 0 added, 1 changed, 1 destroyed.
```

The lifecycle argument also re-orders the graph during a `destroy` operation where all resources are deleted (versus a distribution update). This enables short-lived workspaces to be fully automated.

```console
% terraform destroy
...

Plan: 0 to add, 0 to change, 2 to destroy.
aws_cloudfront_distribution.test: Destroying... [id=E2GSZ0BU4OH406]
aws_cloudfront_distribution.test: Still destroying... [id=E2GSZ0BU4OH406, 00m10s elapsed]
aws_cloudfront_distribution.test: Destruction complete after 16s
aws_cloudfront_response_headers_policy.test: Destroying... [id=7994b2ac-2284-4ff9-86b1-d11974e54a62]
aws_cloudfront_response_headers_policy.test: Destruction complete after 0s

Destroy complete! Resources: 2 destroyed.
```

---

To close this issue out I will be adding an acceptance test for this configuration and updating the registry documentation with an example for using the resources together.
